drivers/regulator/regulator_common.c

```diff
@@ -183,12 +183,16 @@ int regulator_disable(const struct device *dev)
 	(void)k_mutex_lock(&data->lock, K_FOREVER);
 #endif
 
-	data->refcnt--;
+	if (data->refcnt < 1) {
+		ret = -EALREADY;
```

Comment:
> is -EALREADY documented as a valid interface return code?

Can you please elaborate? I've added it to regulator_disable() description and it also had "-errno Negative errno in case of failure." before that, but I believe that's not what you're asking about.

> do we want to return an error code here? maybe just return 0 + assert?

I'm not sure, why that would be better?

Reply:
> > is -EALREADY documented as a valid interface return code?
> 
> Can you please elaborate? I've added it to regulator_disable() description and it also had "-errno Negative errno in case of failure." before that, but I believe that's not what you're asking about.

The last case exists essentially for error propagation, but what an interface is known to return is explicitly documented.

> 
> > do we want to return an error code here? maybe just return 0 + assert?
> 
> I'm not sure, why that would be better?

Such a situation will happen because of unbalanced enables/disables, ie, a programming error.

Reply:
Fair point. Although, there is at least one instance where unbalanced enables/disables could happen without a programming error - regulator shell module.

So, if I make this an assertion instead, application shell user can trigger it as long as regulator shell and asserts are enabled. Unless regulator shell module will have external reference counter for every regulator in system, but I don't think that would be a good solution.

Maybe it would be better to ignore simply ignore excessive disables without returning error and without asserting?

Reply:
If the main concern is regulator shell calls, we could check 'is_enabled' there and print a warning if already disabled.

Reply:
Good idea, but then there will be a potential race condition if there are other parts of code that use regulators:
1. Shell code checks is_enabled() and it returns true
2. Other task disables the regulator
3. Shell calls disable() on it again, triggering the assertion if it is there and is enabled